### PR TITLE
First version

### PR DESCRIPTION
**Pull request by cyrusHwang**

### CODE REVIEW

File: cIDs_txt_files/csv_osu.txt [hunk 1]
```diff
@@ -0,0 +1,148 @@
+2394511
```
Comment (AlexEdgcomb) [inline cIDs_txt_files/csv_osu.txt]: We'll want to document where the files inside the folder `cIDs_txt_files ` came from. Perhaps add a README to `cIDs_txt_files` that says where and how to get these files. This will be useful for months from now when we've forgotten how to use the scripts :)

File: DataAnalysis.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+from lib import misc_funcs, incorrect_subs, csv_downloader, correct_subs, user_data, activity_difficulty
+import os
+
+#email = input("Email: ").strip()
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Please remove unused comments

File: DataAnalysis.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    pass
+elif activityType == '2':
+    choice = input(
+    '''Specification Options
+    1) manual entry of content IDs
+    2) rIDs text file
+    3) manual entry of chapter[.lesson]s
+    4) all programming activities
+Input a number: ''').strip()
+    if choice == '1':
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Each of these choices is heavily redundant. Can you make a pass to refactor?

File: DataAnalysis.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+zyBookCode = input("zyBook code of zyBook of interest: ").strip()
+activityType = input("Enter 1 for progression activities or 2 for programming activities: ").strip()
+
+if activityType == '1':
+    pass
+elif activityType == '2':
+    choice = input(
+    '''Specification Options
+    1) manual entry of content IDs
+    2) rIDs text file
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Would be good to document in the code what rIDs are, or where they come from. Maybe just point the user to the folder with the existing rIDs text files?

File: DataAnalysis.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    3) manual entry of chapter[.lesson]s
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: What is lesson? Do you mean section? Also, please give an example in this output. `Ex: 5.1 for chapter 5, section 1`

File: DataAnalysis.py (hunk 3 above)
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Maybe, for example, move this code up:
```
if choice in ['1', '2', '3', '4']:
    print("Downloading csvs...")
    if not os.path.exists("original_csvs/{}".format(zyBookCode)):
        os.makedirs("original_csvs/{}".format(zyBookCode))
```

File: DataAnalysis.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    '''Data Generations Available
+    1) correct submissions
+    2) incorrect submissions
+    3) user data
+    4) activity difficulty
+Input numbers separated by spaces: ''').strip().split(' ')
+    print("Generating data...")
+    for choice in choices:
+        if choice == '1':
+            if not os.path.exists("generated_csvs/{}/correct_subs".format(zyBookCode)):
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Please refactor redundancy.

Comment (AlexEdgcomb) [inline DataAnalysis.py]: Nvm, in this case.

File: DataAnalysis.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                os.makedirs("generated_csvs/{}/correct_subs".format(zyBookCode))
+            correct_subs.generateData(csvListPaths)
+        elif choice == '2':
+            if not os.path.exists("generated_csvs/{}/incorrect_subs".format(zyBookCode)):
+                os.makedirs("generated_csvs/{}/incorrect_subs".format(zyBookCode))
+            incorrect_subs.generateData(csvListPaths)
+        elif choice == '3':
+            user_data.generateData(csvListPaths)
+        elif choice == '4':
+            activity_difficulty.generateData(csvListPaths)
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Maybe end with a `Done` message?

File: lib/activity_difficulty.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import csv
+from collections import defaultdict
+from lib import misc_funcs
+from statistics import median, mean
+
+def generateData(csvListPaths):
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Please document what this script does

File: DataAnalysis.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+from lib import misc_funcs, incorrect_subs, csv_downloader, correct_subs, user_data, activity_difficulty
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Please document what this script does in a comment

File: lib/activity_difficulty.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            cIdToAttsData[contentID]["avgAtts"] = round(mean(userIdToNumAtts.values()), 2)
+            cIdToAttsData[contentID]["numUsersPart"] = numUsersParticipated
+    sortedData = sorted(cIdToAttsData.items(), key=lambda entry : entry[1]["medianAtts"], reverse=True)
+    outputCsv = open("generated_csvs/{}/{}_activity_difficulty.csv".format(zyBookCode, zyBookCode),
+                     "w", newline='', encoding="utf8")
+    writer = csv.writer(outputCsv)
+    writer.writerow(["Content ID", "Median Number of Attempts", "Average Number of Attempts", "Number of Users Participated"])
+    for cID, data in sortedData:
+        writer.writerow([cID, data["medianAtts"], data["avgAtts"], data["numUsersPart"]])
+
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Can remove blank line

File: lib/correct_subs.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import csv
```
Comment (AlexEdgcomb) [inline lib/correct_subs.py]: Please document what this script does in a comment

File: lib/csv_downloader.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
+from selenium import webdriver
```
Comment (AlexEdgcomb) [inline lib/csv_downloader.py]: Please document what this script does in a comment.

File: lib/csv_downloader.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        responseURL = "https://s2.zybooks.com/v1/zybook/{}/submissions/{}?auth_token={}".format(zyBookCode, contentID, authToken)
+        urlContentText = urllib.request.urlopen(responseURL)
+        urlContentJsonObj = json.loads(urlContentText.read().decode("utf-8-sig"))
+        csvURL = urlContentJsonObj["url"]
+        csvPath = "original_csvs/{}/{}_{}.csv".format(zyBookCode, zyBookCode, contentID)
+        urllib.request.urlretrieve(csvURL, filename=csvPath)
+        csvListPaths.append(csvPath)
+    return csvListPaths
+
+
```
Comment (AlexEdgcomb) [inline lib/csv_downloader.py]: Can remove extra lines

File: lib/csv_downloader.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        urlContentText = urllib.request.urlopen(responseURL)
+        urlContentJsonObj = json.loads(urlContentText.read().decode("utf-8-sig"))
+        csvURL = urlContentJsonObj["url"]    
+        csvPath = "original_csvs/{}/{}_{}.csv".format(zyBookCode, zyBookCode, contentID)
+        urllib.request.urlretrieve(csvURL, filename=csvPath)
+        csvListPaths.append(csvPath)
+    return csvListPaths
+    
+def byChLess(authToken, zyBookCode, activityType, chLessList, email, password):
+    pass
```
Comment (AlexEdgcomb) [inline lib/csv_downloader.py]: Is this supposed to be `pass`? Or just not implemented yet? If not implemented, perhaps add a TODO message. Ex: `print("TODO: Add ....")` where `...` is whatever this is supposed to do

File: lib/csv_downloader.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+def all(authToken, zyBookCode, activityType, email, password):
```
Comment (AlexEdgcomb) [inline lib/csv_downloader.py]: Please document what this does

File: lib/csv_downloader.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+import urllib.request
+import json
+
+def byCIDs(authToken, zyBookCode, contentIDsList):
```
Comment (AlexEdgcomb) [inline lib/csv_downloader.py]: Please document what this does

File: lib/incorrect_subs.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
+import csv
```
Comment (AlexEdgcomb) [inline lib/incorrect_subs.py]: Please document what this script does in a comment

File: lib/incorrect_subs.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+from collections import defaultdict
+from lib import misc_funcs
+
+def generateData(csvListPath):
```
Comment (AlexEdgcomb) [inline lib/incorrect_subs.py]: Please give a more meaningful name

Comment (AlexEdgcomb) [inline lib/incorrect_subs.py]: Also, please document what this function does

File: lib/misc_funcs.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    listPartition = string.split('"')
+    for index, segment in enumerate(listPartition):
+        if index%2 == 0:
+            listPartition[index] = re.sub("\s+", "", segment)
+            
+    return ('"'.join(listPartition))
+
+def getAuthToken(email, password):
+    pass
+
```
Comment (AlexEdgcomb) [inline lib/misc_funcs.py]: Can remove extra line

File: lib/misc_funcs.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    #removing all whitespace except whitespace within quotes
+    listPartition = string.split('"')
+    for index, segment in enumerate(listPartition):
+        if index%2 == 0:
+            listPartition[index] = re.sub("\s+", "", segment)
+            
+    return ('"'.join(listPartition))
+
+def getAuthToken(email, password):
+    pass
```
Comment (AlexEdgcomb) [inline lib/misc_funcs.py]: Supposed to be blank? If not, please add print message

File: lib/misc_funcs.py [hunk 21]
```diff
@@ -0,0 +1,19 @@
+import re
+import requests
+
+def stripNoiseSub(string : str):
```
Comment (AlexEdgcomb) [inline lib/misc_funcs.py]: What's noise sub? Please explain in documentation

File: lib/user_data.py [hunk 22]
```diff
@@ -0,0 +1,58 @@
+import csv
+from collections import defaultdict
+from lib import misc_funcs
+from statistics import median, mean
+
+def generateData(csvListPaths):
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please give a more meaningful name

Comment (AlexEdgcomb) [inline lib/user_data.py]: Also, doc what this does

File: lib/user_data.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
+import csv
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please document what this script does in a comment

File: lib/user_data.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+        for contentID, numAtts in sortedData:
+            print("{}: {}".format(contentID, numAtts))
+        contentID = input("\nEnter content ID that you would like to investigate ('q' to quit): ")
+        if contentID == 'q':
+            break
+        print("\nSubmissions Log")
+        for attemptNum, rawSub in enumerate(userIdToData[userID]["cIdToAtts"][contentID], 1):
+            print("({})\n{}\n".format(attemptNum, rawSub))
+        
+
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: extra lines

File: lib/user_data.py [hunk 25]
```diff
@@ -0,0 +1,58 @@
+import csv
+from collections import defaultdict
+from lib import misc_funcs
+from statistics import median, mean
+
+def generateData(csvListPaths):
+    userIdToData = defaultdict(lambda : {"cIdToNumAtts" : defaultdict(int), "cIdToAtts" : defaultdict(list), "cIdsCompleted" : set()})
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please comment what this line is doing

File: lib/user_data.py [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    writer.writerow(["User ID", "Median Number of Attempts", "Average Number of Attempts",
+                     "Number of Activities Participated", "Number of Activities Completed",
+                     "Content IDs of Unfinished Activities"])
+    for path in csvListPaths:
+        analyzedUsers = set()
+        contentID = path.split('/')[2].split('_')[1].rstrip(".csv")
+        with open(path, 'r', encoding="utf8") as csvFile:
+            dataTable = csv.reader(csvFile)
+            next(dataTable)
+            for dateAndTime, userID, isCorrect, rawSub in dataTable:
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please doc what this loop is doing

File: lib/user_data.py [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+                if userID not in analyzedUsers:
+                    if int(isCorrect):
+                        analyzedUsers.add(userID)
+                        userIdToData[userID]["cIdsCompleted"].add(contentID)
+                    userIdToData[userID]["cIdToAtts"][contentID].append(rawSub)
+                    userIdToData[userID]["cIdToNumAtts"][contentID] += 1
+    for userID in userIdToData:
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please doc what this loop is doing in plain English

File: lib/user_data.py [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+        listNumAtts = userIdToData[userID]["cIdToNumAtts"].values()
+        userIdToData[userID]["numActsPart"] = len(listNumAtts)
+        userIdToData[userID]["numActsCompleted"] = len(userIdToData[userID]["cIdsCompleted"])
+        setCidUnfinished = set(userIdToData[userID]["cIdToNumAtts"].keys()) - userIdToData[userID]["cIdsCompleted"]
+        userIdToData[userID]["cIdsUnfinished"] = ", ".join(setCidUnfinished)
+        userIdToData[userID]["avgNumAtts"] = mean(listNumAtts)
+        userIdToData[userID]["medNumAtts"] = median(listNumAtts)
+    sortedData = sorted(userIdToData.items(),
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please doc the sorting priorities/method

File: lib/user_data.py [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+                        reverse = True)
+    for userID, data in sortedData:
+        writer.writerow([userID, data["medNumAtts"], data["avgNumAtts"], data["numActsPart"],
+                         data["numActsCompleted"], data["cIdsUnfinished"]])
+    outputCsv.close()
+    while True:
+        userID = input("Enter user ID of user of interest ('q' to quit): ")
+        if userID == 'q':
+            break
+        sortedData = sorted(userIdToData[userID]["cIdToNumAtts"].items(), key=lambda entry : entry[1], reverse=True)
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please explain the sorting method/priorities

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , I've made a code review pass. :)

**Comment by cyrusHwang** [comment]
@AlexEdgcomb Ready for review.

File: lib/correct_subs.py (hunk 11 above)
Comment (AlexEdgcomb) [inline lib/correct_subs.py]: This was done

File: lib/incorrect_subs.py (hunk 17 above)
Comment (AlexEdgcomb) [inline lib/incorrect_subs.py]: This was done

File: lib/user_data.py (hunk 23 above)
Comment (AlexEdgcomb) [inline lib/user_data.py]: This was done

File: lib/misc_funcs.py [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+            
+    return ('"'.join(listPartition))
+
+def getAuthToken(email, password):
+    d = {"email":email, "password":password}
+    h = {"Content-type":"application/json", "Accept":"text/plain"}
+    r = requests.post("https://s2.zybooks.com/v1/signin", data=json.dumps(d), headers=h)
+    json_data = r.json()
+    return json_data["session"]["auth_token"]
+
```
Comment (AlexEdgcomb) [inline lib/misc_funcs.py]: There is extra whitespace here

File: original_csvs/.gitignore [hunk 31]
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
Comment (AlexEdgcomb) [inline original_csvs/.gitignore]: Btw, you should be able to have only 1 .gitignore file for the whole repo, then you can add special rules for sub-folders.

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , much better documented! I've added 2 new comments. Looks like there are ~3 old comments that still need to be addressed.

File: DataAnalysis.py (hunk 4 above)
Comment (cyrusHwang) [inline DataAnalysis.py]: Done.

File: DataAnalysis.py (hunk 5 above)
Comment (cyrusHwang) [inline DataAnalysis.py]: Fixed.

File: DataAnalysis.py (hunk 9 above)
Comment (cyrusHwang) [inline DataAnalysis.py]: Done.

**Comment by cyrusHwang** [comment]
@AlexEdgcomb Ready for review.

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , looks pretty good. Have you tested to make sure this runs as expected? Perhaps ask Kevin to test as well? I won't have time to do a deep dive into testing.
